pygdf/queryutils.py

```diff
@@ -0,0 +1,205 @@
+    colnames = sorted(extractor.colnames)
+    refnames = sorted(extractor.refnames)
+    info = {
+        'source': text,
+        'args': colnames + refnames,
+        'colnames': colnames,
+        'refnames': refnames,
+    }
+    return info
+
+
+def query_builder(info, funcid):
+    """Function builder for the query expression
+
+    Parameters
+    ----------
+    info: dict
+        From the `query_parser()`
+    funcid: str
+        The name for the function being generated
+
+    Returns
+    -------
+    func: a python function of the query
+    """
+    args = info['args']
+    def_line = 'def {funcid}({args}):'.format(funcid=funcid,
+                                              args=', '.join(args))
+    lines = [def_line, '    return {}'.format(info['source'])]
+    source = '\n'.join(lines)
+    glbs = {}
+    six.exec_(source, glbs)
+    return glbs[funcid]
+
+
+def _check_error(tree):
+    if not isinstance(tree, ast.Module):
+        raise QuerySyntaxError('top level should be of ast.Module')
+    if len(tree.body) != 1:
+        raise QuerySyntaxError('too many expressions')
+
+
+_cache = {}
+
+
+def query_compile(expr):
+    """Compile the query expression.
+
+    Parameters
+    ----------
+    expr : str
+        The boolean expression
+
+    Returns
+    -------
+    compiled: dict
+        key "kernel" is the cuda kernel for the query.
+        key "args" is a sequence of name of the arguments.
+    """
+
+    funcid = 'queryexpr_{:x}'.format(np.uintp(hash(expr)))
+    compiled = _cache.get(funcid)
+    if compiled is None:
+        info = query_parser(expr)
+        fn = query_builder(info, funcid)
+        args = info['args']
+        # compile
+        devicefn = cuda.jit(device=True)(fn)
+
+        kernelid = 'kernel_{}'.format(funcid)
+        kernel = _wrap_query_expr(kernelid, devicefn, args)
+
+        compiled = info.copy()
+        compiled['kernel'] = kernel
+        _cache[funcid] = compiled
+    return compiled
+
+
+_kernel_source = '''
+@cuda.jit
+def {kernelname}(out, {args}):
+    idx = cuda.grid(1)
+    if idx < out.size:
+        out[idx] = queryfn({indiced_args})
+'''
+
+
+def _wrap_query_expr(name, fn, args):
+    """Wrap the query expression in a cuda kernel.
+    """
+    def _add_idx(arg):
+        if arg.startswith(ENVREF_PREFIX):
+            return arg
+        else:
+            return '{}[idx]'.format(arg)
+
+    def _add_prefix(arg):
+        return '_args_{}'.format(arg)
+
+    glbls = {
+        'queryfn': fn,
+        'cuda': cuda,
+    }
+    kernargs = map(_add_prefix, args)
+    indiced_args = map(_add_prefix, map(_add_idx, args))
+    src = _kernel_source.format(kernelname=name,
+                                args=', '.join(kernargs),
+                                indiced_args=', '.join(indiced_args))
+    six.exec_(src, glbls)
+    kernel = glbls[name]
+    return kernel
+
+
+def query_execute(df, expr, callenv):
+    """Compile & execute the query expression
```

Comment:
is there any value in trying to cache these functions?  I'm not sure what the scheme would be that would allow the same expression to be used with different data frames and also handle changes to the environment...